File: solutions/cpp/knapsack/1/knapsack.cpp
#include "knapsack.h"

namespace knapsack
{
    auto maximum_value(int max_weight, List items) -> int
    {
        if (items.pos == items.end or max_weight == 0) return 0;

        if (items.pos->weight > max_weight)
            return maximum_value(max_weight, {items.pos + 1, items.end});

        return std::max(
                items.pos->value + maximum_value(
                                        max_weight - items.pos->weight,
                                        {items.pos + 1, items.end}),
                maximum_value(max_weight, {items.pos + 1, items.end}));
    }
} // namespace knapsack

File: solutions/haskell/knapsack/1/src/Knapsack.hs
module Knapsack (maximumValue) where

type Weight = Int
type Value  = Int

maximumValue :: Weight -> [(Weight, Value)] -> Value
maximumValue _ [] = 0
maximumValue n ((w, v):xs)
 | w > n = maximumValue n xs
 | otherwise = max 
                (v + maximumValue (n - w) xs) 
                (maximumValue n xs)
